madara/crates/tests/src/lib.rs

```diff
@@ -131,9 +131,15 @@ impl Drop for MadaraCmd {
 
         // Ensure process cleanup
         let _ = child.wait();
+        std::thread::sleep(Duration::from_millis(500));
     }
 }
 
+fn is_port_available(port: u16) -> bool {
+    use std::net::TcpListener;
+    TcpListener::bind(("127.0.0.1", port)).is_ok()
```

Comment:
I think you can do `let socket = TcpListener::bind("127.0.0.1:0").unwrap();` and let the OS find a port for you

Reply:
I did not know this, it's very useful. Completely rewrote the async logic here to take advantage of this fact.